### ISSUE TITLE
[CUDA] MultiheadAttention with masks and dropout produces NaNs

### ISSUE THREAD

**Issue by YChienHung** [pytorch/pytorch#152028]
### 🐛 Describe the bug

I
have
same
problem
with the code below, but it disappeared with `need_weights=True`.So what can I do to solve it if it is a good old contiguous tensor problem ?
```python

self.read_from_pixel = CrossAttention(
    self.embed_dim, self.num_heads, add_pe_to_qkv=this_cfg.read_from_pixel.add_pe_to_qkv
)

# bs*C*H*W -> bs*(H*W)*C
pixel_flat = pixel.flatten(2, 3).transpose(1, 2).contiguous()

# masked cross attention
x, q_weights = self.read_from_pixel(
    x, pixel_flat, query_pe, pixel_pe, attn_mask=attn_mask, need_weights=need_weights
)
```
```python

class CrossAttention(nn.Module):
    def __init__(
            self, dim: int, nhead: int, dropout: float = 0.0, batch_first: bool = True,
            add_pe_to_qkv: List[bool] = [True, True, False], residual: bool = True, norm: bool = True
    ):
        super().__init__()
        self.cross_attn = nn.MultiheadAttention(dim, nhead, dropout=dropout, batch_first=batch_first)
        if norm:
            self.norm = nn.LayerNorm(dim)
        else:
            self.norm = nn.Identity()
        self.dropout = nn.Dropout(dropout)
        self.add_pe_to_qkv = add_pe_to_qkv
        self.residual = residual

    def forward(
            self, x: torch.Tensor, mem: torch.Tensor, x_pe: torch.Tensor, mem_pe: torch.Tensor, attn_mask: bool = None,
            *,
            need_weights: bool = False
    ) -> (torch.Tensor, torch.Tensor):
        x = self.norm(x)
        if self.add_pe_to_qkv[0]:
            q = x + x_pe
        else:
            q = x

        if any(self.add_pe_to_qkv[1:]):
            mem_with_pe = mem + mem_pe
            k = mem_with_pe if self.add_pe_to_qkv[1] else mem
            v = mem_with_pe if self.add_pe_to_qkv[2] else mem
        else:
            k = v = mem
        r = x
        x, weights = self.cross_attn(
            q, k, v, attn_mask=attn_mask, need_weights=need_weights, average_attn_weights=False
        )

        if self.residual:
            return r + self.dropout(x), weights
        else:
            return self.dropout(x), weights


```

### Versions

None

**Comment by kimmeoungjun** [comment]
Hi. I reproduced this on CUDA/A100 with PyTorch 2.8.0+cu128.

The issue is in the `nn.MultiheadAttention(..., need_weights=True)` path. When a bool `attn_mask` fully masks out a query row, that row’s attention scores become all `-inf`; the current regular softmax then produces NaNs. The `need_weights=False` path goes through SDPA and already handles fully masked rows with safe-softmax semantics, returning zeros instead.

I have a fix ready that updates the `need_weights=True` path to use `torch._safe_softmax`, making it consistent with the SDPA path:

```python
attn_output_weights = torch._safe_softmax(attn_output_weights, dim=-1)
```

I also prepared a regression test for `MultiheadAttention` with dropout + bool `attn_mask`, covering both `need_weights=False` and `need_weights=True`, and verifying that fully masked rows produce zero output/weights instead of NaNs.

Any thoughts on this approach before I open a PR? Thank you.

**Comment by eqy** [comment]
PyTorch 2.8 is quite old at this point, could you check if nightlies which contain this commit https://github.com/pytorch/pytorch/pull/177868 still show this issue?

**Comment by kimmeoungjun** [comment]
Thanks for the pointer. I rechecked this on an A100 with clean wheel installs and without my local patch.

I tested:

* `torch==2.11.0+cu128`
* `torch==2.12.0.dev20260408+cu128`

  * `torch.version.git_version == 6b1da8db53541ea06a7d5d6ca508938485bf1660`

On both builds, the issue still reproduces for `nn.MultiheadAttention(..., need_weights=True)`:

```text
# torch==2.11.0+cu128
need_weights=False output_has_nan=False weights_has_nan=None first_row_zero=True first_weight_row_zero=None
need_weights=True  output_has_nan=True  weights_has_nan=True first_row_zero=False first_weight_row_zero=False

# torch==2.12.0.dev20260408+cu128
need_weights=False output_has_nan=False weights_has_nan=None first_row_zero=True first_weight_row_zero=None
need_weights=True  output_has_nan=True  weights_has_nan=True first_row_zero=False first_weight_row_zero=False
```

The repro script is:

```python
import torch
import torch.nn as nn


def run(need_weights: bool):
    device = "cuda"
    torch.manual_seed(0)

    mha = nn.MultiheadAttention(
        8,
        2,
        dropout=0.5,
        batch_first=True,
        bias=False,
        device=device,
    )
    mha.train()

    query = torch.randn(2, 3, 8, device=device)
    key = torch.randn(2, 4, 8, device=device)
    value = torch.randn(2, 4, 8, device=device)

    attn_mask = torch.zeros(3, 4, dtype=torch.bool, device=device)
    attn_mask[0, :] = True

    output, weights = mha(
        query,
        key,
        value,
        attn_mask=attn_mask,
        need_weights=need_weights,
        average_attn_weights=False,
    )

    output_has_nan = torch.isnan(output).any().item()
    weights_has_nan = None if weights is None else torch.isnan(weights).any().item()
    first_row_zero = torch.equal(output[:, 0], torch.zeros_like(output[:, 0]))
    first_weight_row_zero = (
        None
        if weights is None
        else torch.equal(weights[:, :, 0], torch.zeros_like(weights[:, :, 0]))
    )

    print(
        f"need_weights={need_weights} "
        f"output_has_nan={output_has_nan} "
        f"weights_has_nan={weights_has_nan} "
        f"first_row_zero={first_row_zero} "
        f"first_weight_row_zero={first_weight_row_zero}"
    )


print(f"torch={torch.__version__}")
print(f"git_version={getattr(torch.version, 'git_version', None)}")
print(f"cuda_runtime={torch.version.cuda}")
print(f"device={torch.cuda.get_device_name(0)}")
print(f"capability={torch.cuda.get_device_capability(0)}")

for need_weights in (False, True):
    run(need_weights)
```

My understanding is that #177868 is for the SDPA/cuDNN bool `attn_mask` conversion path, while this failing case is the manual `need_weights=True` path in `torch.nn.functional.multi_head_attention_forward`:

```python
attn_output_weights = torch.baddbmm(...)
attn_output_weights = softmax(attn_output_weights, dim=-1)
```

When the bool mask fully masks out a query row, that row’s attention scores become all `-inf`, and regular softmax produces NaNs. The `need_weights=False` path goes through SDPA and already returns zeros for fully masked rows.

The fix I have prepared is to make the `need_weights=True` path use the same safe-softmax semantics:

```python
attn_output_weights = torch._safe_softmax(attn_output_weights, dim=-1)
```

This makes `need_weights=True` consistent with the SDPA path for fully masked rows.

Please let me know if there is a newer nightly or a specific commit you would like me to test before I open the PR.
